src/storage/v2/inmemory/storage.cpp

```diff
@@ -272,111 +272,294 @@ std::optional<VertexAccessor> InMemoryStorage::InMemoryAccessor::FindVertex(Gid
-            "accessor when deleting a vertex!");
-  auto *vertex_ptr = vertex->vertex_;
+  auto res = DetachDelete({vertex}, {}, false);
 
-  std::lock_guard<utils::SpinLock> guard(vertex_ptr->lock);
-
-  if (!PrepareForWrite(&transaction_, vertex_ptr)) return Error::SERIALIZATION_ERROR;
+  if (res.HasError()) {
+    return res.GetError();
+  }
 
-  if (vertex_ptr->deleted) {
+  const auto &value = res.GetValue();
+  if (!value) {
     return std::optional<VertexAccessor>{};
   }
 
-  if (!vertex_ptr->in_edges.empty() || !vertex_ptr->out_edges.empty()) return Error::VERTEX_HAS_EDGES;
+  const auto &[vertices, edges] = *value;
 
-  CreateAndLinkDelta(&transaction_, vertex_ptr, Delta::RecreateObjectTag());
-  vertex_ptr->deleted = true;
+  MG_ASSERT(vertices.size() <= 1, "The number of deleted vertices is not less or equal to 1!");
 
-  // Need to inform the next CollectGarbage call that there are some
-  // non-transactional deletions that need to be collected
-  if (transaction_.storage_mode == StorageMode::IN_MEMORY_ANALYTICAL) {
-    auto *mem_storage = static_cast<InMemoryStorage *>(storage_);
-    mem_storage->gc_full_scan_vertices_delete_ = true;
-  }
-
-  return std::make_optional<VertexAccessor>(vertex_ptr, &transaction_, &storage_->indices_, &storage_->constraints_,
-                                            config_, true);
+  return std::make_optional<VertexAccessor>(vertices[0]);
 }
 
 Result<std::optional<std::pair<VertexAccessor, std::vector<EdgeAccessor>>>>
 InMemoryStorage::InMemoryAccessor::DetachDeleteVertex(VertexAccessor *vertex) {
   using ReturnType = std::pair<VertexAccessor, std::vector<EdgeAccessor>>;
 
-  MG_ASSERT(vertex->transaction_ == &transaction_,
-            "VertexAccessor must be from the same transaction as the storage "
-            "accessor when deleting a vertex!");
-  auto *vertex_ptr = vertex->vertex_;
+  auto res = DetachDelete({vertex}, {}, true);
 
-  std::vector<std::tuple<EdgeTypeId, Vertex *, EdgeRef>> in_edges;
-  std::vector<std::tuple<EdgeTypeId, Vertex *, EdgeRef>> out_edges;
+  if (res.HasError()) {
+    return res.GetError();
+  }
 
-  {
-    std::lock_guard<utils::SpinLock> guard(vertex_ptr->lock);
+  auto &value = res.GetValue();
+  if (!value) {
+    return std::optional<ReturnType>{};
+  }
 
-    if (!PrepareForWrite(&transaction_, vertex_ptr)) return Error::SERIALIZATION_ERROR;
+  auto &[vertices, edges] = *value;
 
-    if (vertex_ptr->deleted) return std::optional<ReturnType>{};
+  MG_ASSERT(vertices.size() <= 1, "The number of detach deleted vertices is not less or equal to 1!");
 
-    in_edges = vertex_ptr->in_edges;
-    out_edges = vertex_ptr->out_edges;
-  }
+  return std::make_optional<ReturnType>(vertices[0], std::move(edges));
+}
+
+Result<std::optional<std::pair<std::vector<VertexAccessor>, std::vector<EdgeAccessor>>>>
+InMemoryStorage::InMemoryAccessor::DetachDelete(std::vector<VertexAccessor *> nodes, std::vector<EdgeAccessor *> edges,
```

Comment:
Basically everything works the same in this situation.  What is really important is that there is no duplication of code for in-memory and disk storage and here you have a bunch of functions with same name